### ISSUE TITLE
A bug of tf.layers.batch_normalization when training is not a constant

### ISSUE THREAD

**Issue by x10000year** [tensorflow/tensorflow#15919]
I encountered a bug of tf.layers.batch_normalization when the training argument is not a constant. Usually, when the training argument evaluates to False, the moving mean and variance should not be updated. However, in certain cases, the moving mean and variance may become NaN.

The bug occurs in the following code in python/layers/normalization.py:

```
    training_value = utils.constant_value(training)
    if training_value is None:
      one_minus_decay = utils.smart_cond(training,
                                         lambda: self._one_minus_decay,
                                         lambda: 0.)
    else:
      one_minus_decay = ops.convert_to_tensor(self._one_minus_decay)
    if training_value or training_value is None:
      mean_update = self._assign_moving_average(self.moving_mean, mean,
                                                one_minus_decay)
      variance_update = self._assign_moving_average(self.moving_variance,
                                                    variance, one_minus_decay)
```

When training is not a constant but evaluates to False, one_minus_decay is set to 0, and it is expected that _assign_moving_average does not actually change the moving average. However, mean and variance are outputs of FusedBatchNormOp, which are not actually computed when training is False. So, the content of mean and variance are random, and it can contain NaN values in certain cases. The NaN values in mean and variance then lead to NaN values in the moving mean and moving variance, even if the one_minus_decay is 0 (NaN times 0 is still NaN). Once moving mean and variance contain NaN values, the model produces NaN outputs forever.

I think a way to fix this issue is to modify _assign_moving_average. Just add following one line: 
`update_delta = tf.cond(tf.equal(one_minus_decay, 0), 0, update_delta)`

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by foges** [comment]
I've run into this issue. Using a python bool for the `training` argument works, but as soon as I replace it with a tf.placeholder I get a `nan` loss. 

Have I written custom code: Yes
OS Platform and Distribution: Ubuntu
TensorFlow installed from: https://www.tensorflow.org/install/install_linux
TensorFlow version: 1.6
Bazel version: N/A
CUDA/cuDNN version: N/A
GPU model and memory: N/A
Exact command to reproduce:
```
  training = tf.placeholder(tf.bool, [])
  ...
  node = tf.layers.batch_normalization(node, training=training)
  ...
  update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS, scope='mynet')
  with tf.control_dependencies(update_ops):
      loss, train_step = loss_fn(label, prediction)
  ...
  # train
  train_loss, _ = sess.run([loss, train_step], feed_dict={training: True, ...})
  print('Train loss: {}'.format(train_loss))  # <- 'Train loss: 0.9273328

  # evaluate
  test_loss = sess.run([loss], feed_dict={training: False, ...})
  print('Test loss: {}'.format(test_loss))  # <-  'Test loss: nan'
```

**Comment by drpngx** [comment]
We're moving to keras in favor of layers. Does this still happen with Keras?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by Kwander** [comment]
@drpngx I can confirm that this does not happen when using Keras layers.

**Comment by drpngx** [comment]
Thanks @Kwander !
